### PR TITLE
Rewrite ODBC connection string parser to mirror msodbcsql ParseAttrStr

### PR DESCRIPTION
**Pull request by saurabh500**
## Summary

[AB#46370](https://sqlclientdrivers.visualstudio.com/b95cf060-8083-439d-8ef1-405d5bf219d8/_workitems/edit/46370)

Replaces the two-phase `tokenize` + interpret connection-string parser in `mssql-odbc/src/connection/mod.rs` with a single-pass, character-by-character **state machine** that reproduces the shipping msodbcsql driver's `ParseAttrStr` (`Sql/Ntdbms/sqlncli/odbc/sqlcconn.cpp`) — **including its quirks**, so apps migrating from `msodbcsql18` see identical parsing semantics.

The public signature `parse_connection_string(&str) -> Result<(ConnectionParams, bool), InvalidAttrValue>` is unchanged, so `api/driver_connect.rs` (which maps `has_warnings` → `01S00` and `Err` → connect failure) needs no change.

## Reproduced msodbcsql behaviors

| Quirk | Behavior |
|---|---|
| Key scan reads **through** `;` | A token without its own `=` merges with following text |
| Missing `=` in remainder | `S_FALSE` — **stop parsing**, keep what was parsed |
| No whitespace trimming | Only leading whitespace/`;` before a key is skipped; `Server =h` no longer matches; value spaces are preserved verbatim |
| `{braced}` values | End at a single `}`; `}}` escapes a literal `}` |
| Junk after `}` | `S_FALSE` — stop (value still stored) |
| Unterminated `{` | `S_FALSE` — swallow rest, stop |
| Recognized-but-unsupported keys | No warning (full `x_rgLookup` set) |
| Unknown keywords | `01S00` warning, **never fail** |
| Invalid value on validated key | `E_FAIL` → `Err` (hard connect failure) |

Diagnostic mapping: `S_OK` → `Ok((_, false))`; `S_FALSE` → `Ok((_, true))` (`01S00` / `SQL_SUCCESS_WITH_INFO`); `E_FAIL` → `Err(InvalidAttrValue)`.

## Notable changes

- Expanded `KNOWN_IGNORED_KEYS` to msodbcsql's recognized-but-not-acted-on set (driver, dsn, app, wsid, applicationintent, multisubnetfailover, columnencryption, tnir, deprecated keys, …) so recognized keys stay silent.
- **Dropped** the ADO.NET/OLE DB-only aliases `Initial Catalog`, `User Id`, and `Password` — the msodbcsql ODBC parser does not recognize them; they are now treated as unknown keys, matching the driver.
- Deliberately **not** mirrored: the fixed-buffer `MAXPATHLEN`/`MAXKEYLEN` length caps (artifacts, not contract).

## Tests

Exhaustive unit tests covering every quirk plus value validation, first-wins duplicates, whitespace fidelity, `}}` escaping, and empty/separator edge cases. `112 passed` for the crate; `cargo fmt --check` and `cargo clippy -- -D warnings` are clean.

## Docs

Adds `mssql-odbc/docs/connection_string_parser.md` documenting the algorithm, the divergence table, the state diagram, and diagnostic mapping.

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

File: mssql-odbc/src/connection/connection_string_parser.rs [hunk 1]
```diff
@@ -0,0 +1,1148 @@
...
+    let mut has_warnings = false;
+
+    loop {
+        // Phase 1 — skip any run of whitespace and ';' before the key. Only the
+        // key's *position* is whitespace-tolerant, not its content. So in
+        // ";;  Server=host" the leading ";;  " is skipped, but in "Server =host"
+        // the trailing space stays part of the key ("server ") and won't match.
+        while i < n && (is_odbc_space(chars[i]) || chars[i] == ';') {
+            i += 1;
+        }
```
[the remark this answers is not in the document]
Comment (saurabh500) [reply mssql-odbc/src/connection/connection_string_parser.rs:405]: Done in 4e28966. Replaced all chars[i] indexing in parse_connection_string with a bounds-checked peek() closure over chars.get(i), so an out-of-range cursor yields None instead of panicking — aligning with the no-panic guideline for this FFI-loaded crate.

File: mssql-odbc/src/connection/mod.rs [hunk 2]
```diff
@@ -1,708 +1,16 @@
...
 //!
-//! Parses `;`-delimited `Key=Value` connection strings per the ODBC spec.
-//! Supports `{braced values}` for values containing `;` or `=`.
-
-use std::fmt;
-
-use crate::connection::odbc_supported_auth_keywords::is_recognized_keyword;
-use tracing::warn;
+//! - [`connection_string_parser`]: parses the `SQLDriverConnect` connection
+//!   string into [`ConnectionParams`], mirroring msodbcsql's `ParseAttrStr`.
```
[the remark this answers is not in the document]
Comment (saurabh500) [reply mssql-odbc/src/connection/mod.rs]: Fixed in 4e28966. The link now uses the qualified path [\ConnectionParams\](connection_string_parser::ConnectionParams) and resolves cleanly (verified with \cargo doc --document-private-items\).

File: mssql-odbc/docs/connection_string_parser.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
+# ODBC Connection String Parser
+
+`mssql-odbc/src/connection/mod.rs` parses the connection string handed to
+`SQLDriverConnect` / `SQLBrowseConnect`. It is a single-pass,
+character-by-character **state machine** that reproduces the behavior of the
+shipping msodbcsql driver's `ParseAttrStr`
+(`Sql/Ntdbms/sqlncli/odbc/sqlcconn.cpp`). The goal is byte-for-byte parity with
+the C++ driver, **including its quirks**, so that applications migrating from
+`msodbcsql18` see identical parsing semantics.
```
[the remark this answers is not in the document]
Comment (saurabh500) [reply mssql-odbc/docs/connection_string_parser.md]: Fixed in 4e28966 — the doc now points at \connection_string_parser.rs\.

File: mssql-odbc/docs/connection_string_parser.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+- The `MAXKEYLEN` (31) key-length cap has no observable effect (no mapped key is
+  that long), so it is omitted.
+
+## Tests
+
+Exhaustive unit tests live in the `#[cfg(test)] mod tests` block of
+`connection/mod.rs` and cover every quirk in the divergence table above plus
+value validation, first-wins duplicates, whitespace fidelity, `}}` escaping,
+separator/empty-input edge cases, and the `S_OK` / `S_FALSE` / `E_FAIL` mapping.
+Run them with `cargo btest` (or `cargo nextest run -p mssql-odbc`).
```
[the remark this answers is not in the document]
Comment (saurabh500) [reply mssql-odbc/docs/connection_string_parser.md:169]: Fixed in 4e28966 — the Tests section now references \connection/connection_string_parser.rs\.

File: mssql-odbc/src/connection/connection_string_parser.rs [hunk 5]
```diff
@@ -0,0 +1,1155 @@
...
+            Some("ActiveDirectoryIntegrated")
+        );
+
+        let (p, ..) =
+            parse_connection_string("Server=h;Trusted_Connection=Yes;Trusted_Connection=No")
+                .unwrap();
+        assert_eq!(p.trusted_connection, Some(true));
+    }
+
+    #[test]
```
Comment (Theekshna) [inline mssql-odbc/src/connection/connection_string_parser.rs:895]: Can we add e2e test(s) for the parser too? 
Also, there is an existing test - `MalformedTokenReturnsSuccessWithInfo`. Currently, this tolerates the behaviour of both msodbcsql & mssql-odbc under the 'KNOWN DIVERGENCE' section. This is because gtests can be run with both the drivers via `./run_e2e.sh --compare-with-msodbcsql`. We should be able to remove the 'known divergence' section now & have it pass with exact msodbcsql behaviour

Comment (saurabh500) [reply mssql-odbc/src/connection/connection_string_parser.rs:895]: Sounds good. Will take care of this

File: mssql-odbc/docs/connection_string_parser.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+| # | Input | Old tokenizer | msodbcsql `ParseAttrStr` (now) |
+|---|-------|---------------|--------------------------------|
+| 1 | `Server=h;bogus;UID=u` | `bogus` skipped, `UID=u` set | key becomes `bogus;UID` (scan reads through `;`); unknown → ignored; **UID never set** |
+| 2 | token with no `=` | skip token, **continue** | `S_FALSE`, **stop parsing** |
+| 3 | `Server =h` (space before `=`) | trimmed → matches `Server` | key = `"Server "` → **no match**, ignored |
+| 4 | `Key= v ` (spaces in value) | trimmed to `v` | value = `" v "` verbatim |
+| 5 | `PWD={a}}b}` | value `a`, rest skipped | value `a}b` (`}}` escape) |
+| 6 | `PWD={val}junk;X=y` | `val`, skip to `;`, continue | `S_FALSE`, **stop** (junk after `}`) |
+| 7 | unterminated `{` | warn, keep partial, continue | `S_FALSE`, stop |
+| 8 | recognized-but-unsupported key (e.g. `QuotedId`) | warns / small allow-list | recognized → **no warning** |
```
Comment (David-Engel) [inline mssql-odbc/docs/connection_string_parser.md:28]: Some of these msodbcsql behaviors seem really bad and lead to negative user experiences. I think this "bogus" example should follow the old tokenizer behavior or produce an error like "Invalid keyword value combination". Seems like msodbcsql behavior diverged from the spec at some point. The ODBC connection string spec clearly says keywords must not contain space, {, or ;.
https://learn.microsoft.com/en-us/openspecs/sql_server_protocols/ms-odbcstr/55953f0e-2d30-4ad4-8e56-b4207e491409
I lean towards taking the mssql-odbc opportunity to fix some of these really bad behaviors.

Comment (saurabh500) [reply mssql-odbc/docs/connection_string_parser.md:28]: I would like to address this change in a follow up PR. 
The scoping question needs to be answered, and I need to give this ask more thought. While I largely agree with erroring out on these weird usecases where we have a key like "bogus" without values etc, there may be other edge cases that may arise. 
Let me know if a follow up works for you.

Comment (David-Engel) [reply mssql-odbc/docs/connection_string_parser.md:28]: A follow up works for me. 👍🏻

File: mssql-odbc/src/connection/connection_string_parser.rs (hunk 5 above)
Comment (saurabh500) [reply mssql-odbc/src/connection/connection_string_parser.rs:895]: Done in dd35deb. I verified both "divergences" empirically by driving `SQLDriverConnect` directly (ctypes against `odbc32.dll`) so I could read the raw return code and **all** diagnostic records against real ODBC Driver 18:

| case | msodbcsql 18 | our parser |
|---|---|---|
| trailing `;;;` / leading / middle | `SUCCESS_WITH_INFO`, **no 01S00** | no warning ✓ |
| buried `noequals` between valid keys | `SQL_ERROR` + `28000` | same ✓ |
| unknown key | `SUCCESS_WITH_INFO` + `01S00` | same ✓ |

**Divergence #2 does not exist** — the claim that msodbcsql posts `01S00` on trailing `;;` is wrong for Driver 18. Its `ParseAttrStr` skips a run of whitespace/`;` at the start of every iteration, so separator runs in any position are consumed cleanly. Our parser already matched, so no parser change was needed. **Divergence #1** is genuinely resolved and now matches exactly.

Changes:
- Removed both KNOWN DIVERGENCE blocks; tightened to exact assertions (`SQL_ERROR`+`28000` for the buried token, no `01S00` for extra semicolons).
- Added `ODBCTestUtils::HasDiagState` that scans **all** diag records. This matters: on a successful login the server posts its own `01000` messages, and the two drivers order the `01S00` warning differently (msodbcsql appends it last → record #3, we post it first). `GetDiagState` only reads record #1, so it would silently miss the warning in `--compare-with-msodbcsql` mode.
- Added a `ConnectionStringParserParityBehaviors` e2e test (braced values, first-wins duplicates, verbatim key matching) plus a unit test locking in separator parity.

Comment (saurabh500) [reply mssql-odbc/src/connection/connection_string_parser.rs:895]: I think the comment about divergence 2 is incorrect. Discovered with e2e tests.

Comment (saurabh500) [reply mssql-odbc/src/connection/connection_string_parser.rs:895]: Correction to my earlier reply: **divergence #2 IS real** — I was wrong, and running the e2e suite in `--compare-with-msodbcsql` mode against a live SQL Server (unixODBC/WSL) caught it. My earlier Windows probe missed it because of the diagnostic **record-ordering** trap: msodbcsql appends the `01S00` after the server's `01000` login messages, and that probe wasn't scanning every record for this case.

Fixed in f19a051.

What actually happens (traced in `ParseAttrStr`, sqlcconn.cpp): after the final value, the parser consumes **exactly one** trailing separator, then the outer loop re-checks `cchAttrStr && *lpsz` (line 4299). So:

| tail | msodbcsql 18 | our parser (after f19a051) |
|---|---|---|
| value at end / single trailing `;` | no `01S00` | no warning ✓ |
| trailing run of **2+** `;` (e.g. `;;`, `;;;`) | `01S00` | warns ✓ |
| trailing `;` + whitespace (e.g. `; `) | `01S00` | warns ✓ |
| leading / between-key runs | no `01S00` | no warning ✓ |

A run of 2+ trailing separators starts a fresh iteration whose leading-skip lands on end-of-input, leaving a degenerate empty key with no `=` → `S_FALSE`/`01S00`. Our parser was silently swallowing all trailing separators; the fix moves the clean end-of-input guard to *before* the leading-skip loop (mirroring line 4299) so trailing runs fall into the empty-key warning path.

Verified with a unixODBC ctypes probe (isolating each case: baseline/`;`/`;;;`/leading/middle/`; `) and the e2e parity harness: **13/13 tests now pass identically on both our driver and real msodbcsql** (previously `driver_connect_test` diverged). Updated the trailing-separator unit tests, the `MalformedTokenReturnsSuccessWithInfo` e2e case, and the parser docs (algorithm, state diagram, stop-conditions).

File: mssql-odbc/src/connection/connection_string_parser.rs [hunk 7]
```diff
@@ -0,0 +1,1228 @@
...
+        let err =
+            parse_connection_string("Server=h;UID=u;PWD=p;Authentication=NotReal").unwrap_err();
+        assert_eq!(err.key, "authentication");
+        assert_eq!(err.value, "NotReal");
+    }
+
+    #[test]
+    fn authentication_managed_identity_is_rejected() {
+        // ActiveDirectoryManagedIdentity is NOT an ODBC keyword — msodbcsql only
+        // accepts ActiveDirectoryMSI (dlgattr.h), which maps to the same method.
```
Comment (David-Engel) [inline mssql-odbc/src/connection/connection_string_parser.rs:827]: I hate that we are perpetuating this mistake. 😢 

No need to change it here since the PR is intentionally replicating msodbcsql behavior. Just sharing disappointment. Maybe we can fix it in the future...

File: mssql-odbc/docs/connection_string_parser.md (hunk 6 above)
Comment (saurabh500) [reply mssql-odbc/docs/connection_string_parser.md:28]: https://github.com/microsoft/mssql-rs/issues/114

File: mssql-odbc/src/connection/connection_string_parser.rs (hunk 7 above)
Comment (saurabh500) [reply mssql-odbc/src/connection/connection_string_parser.rs:827]: What's the ask? `ActiveDirectoryManagedIdentity` should be considered valid in addition to `ActiveDirectoryMSI`?

Comment (David-Engel) [reply mssql-odbc/src/connection/connection_string_parser.rs:827]: At least, yes. Background is "MSI" was the Azure "Managed Service Identity" feature before it was renamed to just "Managed Identity". ODBC released with the acronym right before the feature rename and because the team was all for keeping connection strings short (when we thought connection string length had a 1k limit). The other drivers all have ActiveDirectoryManagedIdentity and ActiveDirectoryMSI as an alias for conn string compatibility.

Comment (saurabh500) [reply mssql-odbc/src/connection/connection_string_parser.rs:827]: https://github.com/microsoft/mssql-rs/issues/115 Tracking item

**Comment by David-Engel** [comment]
Suggestion: the new parser-parity e2e tests build their own connection strings from `cfg.Server()`, `cfg.Uid()`, and `cfg.Pwd()`, but the `DriverConnectLiveTest` fixture only checks `HasConnection()`. That also allows valid e2e configurations that use `ODBC_TEST_CONNSTR`, DSN, or integrated auth, where those individual SQL-auth fields can be empty. In those environments these tests can fail even though the suite is configured correctly.

Could we either skip the parser-parity tests unless `Server`, `UID`, and `PWD` are all present, or add a small helper/fixture for tests that explicitly require mutable SQL-auth field configuration?

**Comment by saurabh500** [comment]
@ttk what's your advise on @David-Engel 's comment about e2e testing ?

**Comment by saurabh500** [comment]
Two things:

Tactical fix (in this PR, commit f457cff): added `ODBCTestConfig::HasSqlAuth()` (Server && UID && PWD all present) and `GTEST_SKIP()` at the top of the two parser-parity tests (`MalformedTokenReturnsSuccessWithInfo`, `ConnectionStringParserParityBehaviors`). They now skip cleanly in `ODBC_TEST_CONNSTR`, DSN, and integrated-auth environments instead of failing on an empty-credential string. Verified in WSL/unixODBC: with SQL auth the full suite is 13/13; without UID/PWD the two tests report SKIPPED (0 failures).

Proper design (follow-up #117): a capability model on the test config (`enum class ODBCCap { SqlAuth, MutableDatabase, IntegratedAuth, Dsn, Encrypt }` + `HasCapability()` + a one-line `ODBC_REQUIRE_CAP(...)` skip helper), plus a narrow `BuildConnectionStringNarrow()` so auth-neutral parser cases append to the already-configured connection rather than rebuilding it. That lets "any connection that authenticates" tests run everywhere, while tests that need mutable SQL-auth fields declare it explicitly.

@ttk and @david-engel since I have to automate the runs of these tests in the pipeline, I have created an issue to track the management of configurations needed for various tests. For this PR I have incorporated the fix that David suggested.

**Review by Theekshna — APPROVED** [review]
